### PR TITLE
Add TypeScript typings to project

### PR DESCRIPTION
**Pull request by blakeembrey**
Closes https://github.com/kevinbeaty/any-promise/issues/18. I can add this as the first project to https://github.com/typings/typings/issues/480, when I get to it 😄 Hopefully that'll always give you somewhere you can get help, just in case. But you can always ping me with changes also, it's pretty easy to pick up.

### CODE REVIEW

File: register.d.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import Promise = require('./index');
+
+declare interface Register {
+  Promise: typeof Promise;
+  implementation: string;
+}
+
+declare function register (module?: string): Register;
```
Comment (kevinbeaty) [inline register.d.ts]: `register` now has an options object that can contain either a `Promise` constructor or `global` flag, so would this be something like: 

```typescript
declare interface RegisterOptions {
    Promise?: typeof Promise;
    global?: boolean
}
 declare function register (module?: string, RegisterOptions?): Register; 
```

I completely made that up based on this PR alone.  If I got anywhere close to right, yay for TypeScript!! ;)

**Comment by kevinbeaty** [comment]
Cool, thanks.

A couple more things.  #19 added "registration shortcuts".  These are just bare imports that don't export anything, e.g.:

```js
require('any-promise/register/bluebird')

// is shorthand for
require('any-promise/register')('bluebird', {Promise: require('bluebird'))
```

Would there be any reason to add definitions for these shortcuts?

Also, #15 (optional import) is probably coming soon in some form.  The interface is still being discussed so I'll probably bug you again when that lands.

Thanks again!

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  static all <T1, T2, T3, T4, T5, T6> (values: [T1 | Promise.Thenable<T1>, T2 | Promise.Thenable<T2>, T3 | Promise.Thenable<T3>, T4 | Promise.Thenable <T4>, T5 | Promise.Thenable<T5>, T6 | Promise.Thenable<T6>]): Promise<[T1, T2, T3, T4, T5, T6]>;
+  static all <T1, T2, T3, T4, T5> (values: [T1 | Promise.Thenable<T1>, T2 | Promise.Thenable<T2>, T3 | Promise.Thenable<T3>, T4 | Promise.Thenable <T4>, T5 | Promise.Thenable<T5>]): Promise<[T1, T2, T3, T4, T5]>;
+  static all <T1, T2, T3, T4> (values: [T1 | Promise.Thenable<T1>, T2 | Promise.Thenable<T2>, T3 | Promise.Thenable<T3>, T4 | Promise.Thenable <T4>]): Promise<[T1, T2, T3, T4]>;
+  static all <T1, T2, T3> (values: [T1 | Promise.Thenable<T1>, T2 | Promise.Thenable<T2>, T3 | Promise.Thenable<T3>]): Promise<[T1, T2, T3]>;
+  static all <T1, T2> (values: [T1 | Promise.Thenable<T1>, T2 | Promise.Thenable<T2>]): Promise<[T1, T2]>;
+  static all <T1> (values: [T1 | Promise.Thenable<T1>]): Promise<[T1]>;
+  static all <TAll> (values: Array<TAll | Promise.Thenable<TAll>>): Promise<TAll[]>;
+
```
Comment (kevinbeaty) [inline index.d.ts]: Are these multiple declarations necessary for TypeScript or is it an optimization?  It's fine, just curious.

Comment (blakeembrey) [inline index.d.ts]: Yes, required. It allows `Promise.all([a, b]).then` to have proper types. TypeScript doesn't have a concept of variadic types (yet), so we need to manually overload for now.

File: register.d.ts (hunk 1 above)
Comment (blakeembrey) [inline register.d.ts]: Looks correct to me, I'll make the changes 👍

**Comment by blakeembrey** [comment]
@kevinbeaty Committed the changes. On the registration shortcuts, I've copied the plain `Promise` re-exporting. If, at some point, each promise implementation actually adds type definitions, you could import those and re-export those instead.

**Comment by kevinbeaty** [comment]
>  On the registration shortcuts, I've copied the plain Promise re-exporting.

I'm a bit confused now.  Those shortcuts are used only for the side-effect of `register`, nothing is exported.   Why do we declare it as exporting the `Promise`?  (Or maybe it was just a misunderstanding?)

**Comment by blakeembrey** [comment]
@kevinbeaty Oh, I read it wrong - I thought it was exporting the promises directly. Let me double check.

**Comment by blakeembrey** [comment]
@kevinbeaty Switched it to empty exports so that the ES6 import will work (E.g. `import 'any-promise/register/bluebird'`). Module augmentation with global doesn't currently work, so it won't magically appear in the global with TypeScript (yet).

**Comment by kevinbeaty** [comment]
Cool thanks.  Do you recommend anything to help test definitions are correct?  If not, I can take your word for it on this one, but it certainly would be nice.

**Comment by blakeembrey** [comment]
Only using the TypeScript compiler itself, that would ensure they're correct (but not functionally correct).

**Comment by blakeembrey** [comment]
Awesome, cheers! Looking forward to this release, I have a few modules to update now 😄

**Comment by kevinbeaty** [comment]
Published in v 1.3.0

**Pull request merged**